File: easy/remove_parantheses/remove_parantheses.go
package main

import "fmt"

func removeOuterParentheses(S string) string {
	incompleteParanthese := 0
	finalStr := ""
	start := 1

	for i := 0; i < len(S); i++ {
		if S[i] == '(' {
			incompleteParanthese++
		} else {
			incompleteParanthese--
		}
		if incompleteParanthese == 0 {
			end := i
			finalStr = finalStr + string(S[start:end])
			start = i + 2
		}
	}

	return finalStr
}

func main() {
	fmt.Println("Solution 1 : ", removeOuterParentheses("(()())(())"))
	fmt.Println("Solution 2 : ", removeOuterParentheses("(()())(())(()(()))"))
	fmt.Println("Solution 3 : ", removeOuterParentheses("()()"))
}

File: easy/array/remove_parantheses/remove_parantheses.py
class Solution(object):
    def removeOuterParentheses(self, S):
        """
        :type S: str
        :rtype: str
        """
        incomplete_paranthese = 0
        start = 1
        final_string = ""

        for i in range(0, len(S)):
            if S[i] == '(':
                incomplete_paranthese += 1
            else:
                incomplete_paranthese -= 1
            if incomplete_paranthese == 0:
                end = i
                final_string = final_string + S[start:end]
                start = i + 2

        return final_string


s = Solution()
print("Solution 1 :", s.removeOuterParentheses("(()())(())"))
print("Solution 2 :", s.removeOuterParentheses("(()())(())(()(()))"))
print("Solution 3 :", s.removeOuterParentheses("()()"))
